### PR TITLE
support Components

### PR DESCRIPTION
**Pull request by ldanilek**
each component gets its own DatabaseFake.

implement nested transactions with writes for each level of the transaction so they can be rolled back if the inner mutation errors. Nested transactions are not isolated from each other, since I couldn't figure out how to do that. Only outer transactions are isolated.

Also implement function handles as static strings of the form `function://${componentPath};${udfPath}`

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: convex/component.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+});
+
+export const schedule = internalMutation({
+  args: {},
+  handler: async (ctx, _args) => {
+    await ctx.runMutation(components.counter.public.schedule, { name: "beans" });
+  },
+});
+
+// Note this must be a mutation because `createFunctionHandle` writes to the
```
Comment (ldanilek) [inline convex/component.ts:42]: comment is wrong

File: convex/components.test.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    "counter",
+    counterSchema,
+    counterModules
+  );
+  return t;
+}
+
+test("generated attributes", async () => {
+  const t = testWithCounter();
+  const x = await t.mutation(internal.component.directCall);
```
Comment (ldanilek) [inline convex/components.test.ts:23]: rename from directCall and directCall2 and x and test name

File: convex/components.test.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  // const x = await t.query(internal.components.directCall2);
+  expect(x).toEqual(3);
+});
+
+test("component scheduler", async () => {
+  vi.useFakeTimers();
+  const t = testWithCounter();
+  await t.mutation(internal.component.schedule);
```
Comment (ldanilek) [inline convex/components.test.ts:31]: also run query after

File: convex/component.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  args: {},
+  handler: async (ctx, _args) => {
+    const count = await ctx.runQuery(components.counter.public.count, {
+      name: "beans",
+    });
+    return count;
+  },
+});
+
+export const schedule = internalMutation({
```
Comment (ldanilek) [inline convex/component.ts:35]: rename to make it clear this is calling component's mutation that schedules mutation's own function

File: index.ts [hunk 5]
```diff
@@ -112,17 +113,17 @@ class DatabaseFake {
   private _queryResults: Record<QueryId, Array<GenericDocument>> = {};
   // TODO: Make this more robust and cleaner
   jobListener: (jobId: string) => void = () => {};
+  // Pending writes for each level of transaction nesting.
+  // Whenever a child mutation begins, a new level is created for all
+  // components. When a child mutation commits, the writes are merged up one
+  // level. When the top-level mutation commits, the writes are applied to the
+  // database. When a mutation is rolled back, last level of writes is
+  // discarded.
   private _writes: Record<
```
Comment (sshader) [inline index.ts:122]: nit: `Array<Record<...>>`

File: index.ts [hunk 6]
```diff
@@ -1444,6 +1555,146 @@ function findModulesRoot(modulesPaths: string[], userProvidedModules: boolean) {
...
+  // can run mutations in parallel.
+  private _waitOnCurrentFunction: Promise<void> | null = null;
+  private _markTransactionDone: (() => void) | null = null;
+  public functionStack: FunctionPath[] = [];
+
+  async begin(
+    functionPath: FunctionPath,
+    isNested: boolean,
+  ) {
+    if (!isNested) {
```
Comment (sshader) [inline index.ts:1636]: Can we add comment on how we don't have proper transaction serialization for nested calls?

File: index.ts [hunk 7]
```diff
@@ -1444,6 +1555,146 @@ function findModulesRoot(modulesPaths: string[], userProvidedModules: boolean) {
...
+  }
+  "aggregate/ratelimiter"
+}
+
+t.registerComponent("aggregate", schema, glob for files with functions)
+t.registerComponentFake("aggregate", AggregateFake)
+
+subtransaction -- two functions where second errors (and is caught), we don't commit either because we reset writes
+
+*/
```
Comment (sshader) [inline index.ts:1587]: pls clean up sarah's comment lol

File: index.ts [hunk 8]
```diff
@@ -149,34 +153,13 @@ class DatabaseFake {
...
-  endTransaction(token: () => void) {
-    token();
-    this._waitOnCurrentFunction = null;
-  }
-
-  // Used to distinguish between mutation and action execution
-  // environment.
-  isInTransaction() {
-    return this._waitOnCurrentFunction !== null;
+  startTransaction() {
```
Comment (sshader) [inline index.ts:161]: Add comment pointing at `TransactionManager`?

File: index.ts [hunk 9]
```diff
@@ -229,7 +214,10 @@ class DatabaseFake {
     const _creationTime =
       now <= this._lastCreationTime ? this._lastCreationTime + 0.001 : now;
     this._lastCreationTime = _creationTime;
-    this._writes[_id] = {
+    if (!this._writes.length) {
```
Comment (sshader) [inline index.ts:217]:
```suggestion
    if (this._writes.length === 0) {
```

Comment (sshader) [reply index.ts:217]: Maybe make a private method that's like `_addWrite(id, update)` (better name pls)

File: index.ts [hunk 10]
```diff
@@ -313,22 +304,31 @@ class DatabaseFake {
...
+        this._writes[this._writes.length - 1][_id] = write;
       }
     }
-    this.resetWrites();
   }
 
-  resetWrites() {
-    this._writes = {};
+  rollbackWrites() {
+    this._writes.pop();
```
Comment (sshader) [inline index.ts:331]: nit: maybe validate that `_writes` isn't empty?

File: index.ts [hunk 11]
```diff
@@ -407,17 +407,22 @@ class DatabaseFake {
     tableName: string,
     callback: (doc: GenericDocument) => void,
   ) {
-    for (const write of Object.values(this._writes)) {
-      if (write.isInsert) {
-        const document = write.newValue;
-        if (tableNameFromId(document._id) === tableName) {
-          callback(document);
+    for (let i = 0; i < this._writes.length; i++) {
```
Comment (sshader) [inline index.ts:410]: Seems like there's a bug around if there's an insert at one level and a delete / update at a lower level. Regression test + fix?

File: index.ts [hunk 12]
```diff
@@ -407,17 +407,22 @@ class DatabaseFake {
...
+          if (tableNameFromId(document._id) === tableName) {
+            callback(document);
+          }
         }
       }
     }
     for (const document of Object.values(this._documents)) {
       if (tableNameFromId(document._id) === tableName) {
-        const write = this._writes[document._id];
+        let write = undefined;
```
Comment (sshader) [inline index.ts:422]: Can this be `getForWrite`?

File: index.ts [hunk 13]
```diff
@@ -1090,22 +1101,85 @@ function asyncSyscallImpl(db: DatabaseFake) {
...
+          args: udfArgs,
+        } = args;
+        const functionPath = await getFunctionPathFromAddress({
+          name,
+          reference,
+          functionHandle,
+        });
+        if (udfType === "query") {
+          return JSON.stringify(
+            convexToJson(await withAuth().queryFromPath(functionPath, true, udfArgs)),
```
Comment (sshader) [inline index.ts:1119]:
```suggestion
            convexToJson(await withAuth().queryFromPath(functionPath, /* isNested */ true, udfArgs)),
```

File: index.ts [hunk 14]
```diff
@@ -1090,22 +1101,85 @@ function asyncSyscallImpl(db: DatabaseFake) {
...
+          );
+        }
+        if (udfType === "mutation") {
+          return JSON.stringify(
+            convexToJson(
+              await withAuth().mutationFromPath(functionPath, true, udfArgs),
+            ),
+          );
+        }
+        if (udfType === "action") {
```
Comment (sshader) [inline index.ts:1129]: q: does action actually need to be here?

File: index.ts [hunk 15]
```diff
@@ -1090,22 +1101,85 @@ function asyncSyscallImpl(db: DatabaseFake) {
...
+          name,
+          reference,
+          functionHandle,
+        } = args;
+        const functionPath = await getFunctionPathFromAddress({
+          name,
+          reference,
+          functionHandle,
+        });
+        const handle = `function://${functionPath.componentPath};${functionPath.udfPath}`;
```
Comment (sshader) [inline index.ts:1151]: Can we add a comment about the format we're using here (or maybe even add two methods next to each other for creating + parsing function handle?)

File: index.ts [hunk 16]
```diff
@@ -1460,14 +1711,19 @@ export const convexTest = <Schema extends GenericSchema>(
...
 ): TestConvex<SchemaDefinition<Schema, boolean>> => {
-  const db = new DatabaseFake(schema ?? null);
-  (global as unknown as { Convex: any }).Convex = {
-    syscall: syscallImpl(db),
-    asyncSyscall: asyncSyscallImpl(db),
-    jsSyscall: jsSyscallImpl(db),
-    db,
-    modules: moduleCache(modules),
-  };
+  const rootDb = new DatabaseFake(schema ?? null, "");
```
Comment (sshader) [inline index.ts:1714]:
```suggestion
  const rootDb = new DatabaseFake(schema ?? null, /* componentPath */ "");
```
(or `const ROOT_COMPONENT_PATH = ""`)

**Review by sshader — COMMENTED** [review]
More testing ideas:
* Things around multiple layers of transactions
* Things with error handling + rollbacks
* ...can we try writing a test for code that uses a packaged component?

**Comment by ldanilek** [comment]
i'll do this soon

> ...can we try writing a test for code that uses a packaged component

but i think this PR is good to land as-is. it's an improvement on the current state of things, allowing us to at least write test in example apps within component repos

**Comment by ldanilek** [comment]
turns out it does work, although it's more than hacky

```typescript
/// <reference types="vite/client" />

import { expect, test } from "vitest";
import { convexTest } from "convex-test";
import { api, internal } from "./_generated/api";
import schema from "./schema";
import aggregateSchema from "../node_modules/@convex-dev/aggregate/dist/esm/component/schema";

const modules = import.meta.glob("../convex/**/*.*s");
const aggregateModules = import.meta.glob("../node_modules/@convex-dev/aggregate/dist/esm/component/**/*.*s");

test("action calling query", async () => {
  const t = convexTest(schema, modules);
  t.registerComponent("aggregate", aggregateSchema, aggregateModules);
  await t.mutation(api.messages.send, { body: "foo", author: "test" });
  const c = await t.query(api.messages.count);
  expect(c).toStrictEqual(1);
});

```

**Pull request merged**